src/GraphConcept.java

```diff
@@ -0,0 +1,557 @@
+        int[] previous = new int[this.vertices.size()];
+        boolean[] visited = new boolean[this.vertices.size()];
+        Arrays.fill(distance, Integer.MAX_VALUE);
+        Arrays.fill(previous, -1);
+
+        distance[0] = 0;
+
+        for (int i = 1; i <= this.vertices.size(); i++) {
+            int current = -1;
+            int min = Integer.MAX_VALUE;
+
+            for (int j = 0; j < this.vertices.size(); j++) {
+                if (!visited[j] && distance[j] < min) {
+                    current = j;
+                    min = distance[j];
+                }
+            }
+
+            if (current == -1) {
+                break;
+            }
+
+            visited[current] = true;
+
+            for (int neighbor : this.vertices.get(current).neighbors.keySet()) {
+                int alt = this.vertices.get(current).neighbors.get(neighbor);
+
+                if (alt < distance[neighbor]) {
+                    distance[neighbor] = alt;
+                    previous[neighbor] = current;
+                }
+            }
+        }
+
+        ArrayList<Integer> result = new ArrayList<>();
+        for (int at = 0; at < this.vertices.size(); at++) {
+            if (previous[at] != -1) {
+                result.add(at);
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * Representation of a single vertex in the graph.
+     */
+    public static final class Vertex {
+
+        /**
+         * The tag of the vertex.
+         */
+        private int tag;
+
+        /**
+         * The set of tags of the vertex's neighbors.
+         */
+        private Map<Integer, Integer> neighbors;
+
+        /**
+         * Get the tag of the vertex.
+         *
+         * @return the tag of the vertex.
+         */
+        public int getTag() {
+            return this.tag;
+        }
+
+        /**
+         * Constructs a new Vertex object.
+         *
+         * @param tag
+         *            the tag of the vertex
+         */
+        public Vertex(int tag) {
+            this.tag = tag;
+            this.neighbors = new HashMap<Integer, Integer>();
+        }
+
+        /**
+         * Returns if two objects are the same vertex.
+         *
+         * @param obj
+         *            the object to be compared
+         *
+         * @return true if the two objects are the same vertex, false otherwise.
+         */
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj == null || this.getClass() != obj.getClass()) {
+                return false;
+            }
+            Vertex vertex = (Vertex) obj;
+            return this.tag == vertex.tag;
+        }
+
+        /**
+         * Returns the hash code of the vertex.
+         *
+         * @return the hash code of the vertex.
+         */
+        @Override
+        public int hashCode() {
+            return this.tag;
+        }
+    }
```

Comment:
Excellent job! Only thing missing is the main method, but this turned out awesome. Probably more work than needed.